### PR TITLE
[REVIEW] issue244 delegatecall

### PR DESCRIPTION
**Pull request by benjaminbollen**
this is for communication, review and testing purposes only.  Do not merge this into master, until we have properly reviewed and tested

### CODE REVIEW

**Comment by benjaminbollen** [comment]
so reasoning is:
- act on storage of `callee`
- preserve `caller`
- run code from destination
- do not transfer value to destination

**Comment by VoR0220** [comment]
Did you look at this yet? https://github.com/ethereum/go-ethereum/search?utf8=%E2%9C%93&q=delegatecall This has everywhere that is delegate call related in some way and could be an excellent reference point.

File: Godeps/_workspace/src/github.com/tendermint/tendermint/vm/vm.go [hunk 1]
```diff
@@ -147,6 +147,32 @@ func (vm *VM) Call(caller, callee *Account, code, input []byte, value int64, gas
 	return
 }
 
+
+// [This implementation is up for review].
+// DelegateCall is executed by the DELEGATECALL opcode, introduced as off Ethereum Homestead.
+// The intent of delegate call is to run the code of the callee in the storage context of the caller;
+// while preserving the original caller to the previous callee.
+// [ -->] Different to the normal CALL or CALLCODE, the value does not need to be transferred to the callee. [<-- CORRECT?]
```
Comment (VoR0220) [inline Godeps/_workspace/src/github.com/tendermint/tendermint/vm/vm.go]: Not entirely correct. See here. https://github.com/ethereum/go-ethereum/blob/b7e3dfc5a2bc7e2f4d653fbe0ec9774277a10643/core/execution.go#L41

**Comment by silasdavis** [comment]
I'm going to state my understanding of this operation and see if other people agree. "Caller" and "Callee" can get a bit confusing because we an intermediate account (named B below; the delegate) that is both caller and callee.

Suppose we have 3 accounts: A, B, and C with the following psuedo-code contracts attached:

A: 
```
int n := 0
address addr := nil

CALL B
```
B: 
```
int n := 0
address addr := nil

DELEGATECALL C
```
C:
```
int n := 0
address addr := nil

n++
addr = msg.sender
```

Then the result of invoking A (message or transaction) should be the following sequence:

| Stack | A | B | C |
|---------|----|----|----|
| []  | n = 0, addr = nil  | n = 0, addr = nil  | n = 0, addr = nil  |
| [CALL B]  | n = 0, addr = nil  | n = 0, addr = nil  | n = 0, addr = nil  |
| [DELEGATECALL C, CALL B]  | n = 0, addr =nil  | n = 1, addr = A  | n = 0, addr = nil  |

That is B acts a 'delegate' for A in terms of the message sender and value, but like with `CALLCODE` it is B's storage that gets updated by the code in C.

Is this correct?

**Comment by silasdavis** [comment]
I need clarification on what this means, and if it is indeed correct...

>`// [ -->] Different to the normal CALL or CALLCODE, the value does not need to be transferred to the callee. [<-- CORRECT?]`

**Comment by benjaminbollen** [comment]
> That is B acts a 'delegate' for A in terms of the message sender and value, but like with CALLCODE it is > B's storage that gets updated by the code in C.
> Is this correct?
yes, that is also how i understand it and the pseudo-code example is consistent.

**Comment by benjaminbollen** [comment]
> // [ -->] Different to the normal CALL or CALLCODE, the value does not need to be transferred to the > callee. [<-- CORRECT?]
> I need clarification on what this means, and if it is indeed correct...

in the code, opcodes `CALL` and `CALLCODE` will start a new stack and memory by calling the funnction `Call`, which atomically transfers the value from caller to callee; then calls the function `call` which executes the ethereum opcodes;

I understand that the value sent with the `message` (starting a new stack+memory) does not require the value sent from `A` to `B` to be passed to the delegated contract `C`; a few notes:
- whether or not to pass on this value from B to C; does not affect the succes of e-pm test `app35`
- go-ethereum code - from my reading - does also not pass this value to the delegated contract `C`

**Comment by compleatang** [comment]
the key diff between CALL and DELEGATECALL is the storage and memory zone in which the code operates. namely CALL will execute in the storage and memory zone of the CALLEE while DELEGATECALL will execute the in the storage and memory zone of the CALLER.

**Comment by benjaminbollen** [comment]
so in the proposed impl here; for `switch CALL, CALLCODE, DELEGATECALL` it essential hinges on this: 
```
if op == CALLCODE {
    [...]
    ret, err = vm.Call(callee, callee, acc.Code, args, value, gas)
} else if op == DELEGATECALL {
    [...]				
    ret, err = vm.DelegateCall(caller, callee, acc.Code, args, value, gas)
} else {
    [...]
    ret, err = vm.Call(callee, acc, acc.Code, args, value, gas)
}
```
https://github.com/eris-ltd/eris-db/pull/247/files#diff-cad6213f6a08c87e4110c4006789e2b7R832

**Comment by compleatang** [comment]
I'm pretty sure there's a few nuances in the proposed `vm.DelegateCall` which are not captured in geths -> https://github.com/ethereum/go-ethereum/blob/b7e3dfc5a2bc7e2f4d653fbe0ec9774277a10643/core/execution.go#L138-L169

**Comment by VoR0220** [comment]
The original EIP with all the specifications. https://github.com/ethereum/EIPs/issues/23

**Comment by silasdavis** [comment]
I just want to emphasise the difference between `DELEGATECALL` and `CALLCODE` in the example I gave before. I think the following execution would occur with `CALLCODE`: 

| Stack | A | B | C |
|---------|----|----|----|
| []  | n = 0, addr = nil  | n = 0, addr = nil  | n = 0, addr = nil  |
| [CALL B]  | n = 0, addr = nil  | n = 0, addr = nil  | n = 0, addr = nil  |
| [CALLCODE C, CALL B]  | n = 0, addr = nil  | n = 1, addr = **B**  | n = 0, addr = nil  |

Do we agree?

**Comment by benjaminbollen** [comment]
Yes, I agree that the stored address would be now stored as `B` as the message sender.

File: version/version.go [hunk 2]
```diff
@@ -19,4 +19,4 @@ package version
 const TENDERMINT_VERSION = "0.5.0"
 // IMPORTANT: Eris-DB version must be on the last line of this file for
 // the deployment script DOCKER/build.sh to pick up the right label.
-const VERSION = "0.12.0"
+const VERSION = "0.12.0-rc2"
```
Comment (pietv) [inline version/version.go]: @benjaminbollen it's "0.12.0" for #master. without rc.

File: Godeps/_workspace/src/github.com/tendermint/tendermint/vm/vm.go (hunk 1 above)
Comment (VoR0220) [inline Godeps/_workspace/src/github.com/tendermint/tendermint/vm/vm.go]: Just change these notes here please @benjaminbollen then it should be ready to merge.

File: Godeps/_workspace/src/github.com/tendermint/tendermint/vm/vm.go [hunk 3]
```diff
@@ -745,12 +769,21 @@ func (vm *VM) call(caller, callee *Account, code, input []byte, value int64, gas
...
 				return nil, ErrPermission{"call"}
 			}
 			gasLimit := stack.Pop64()
-			addr, value := stack.Pop(), stack.Pop64()
+			addr := stack.Pop()
+			// NOTE: for DELEGATECALL value is preserved from the original
+			// caller, as such it is not stored on stack for the address
+			// to be called; for CALL and CALLCODE value is stored on stack
+			// and needs to be overwritten from the given value.
+			// TODO: [ben] assert that malicious code cannot produce
```
Comment (silasdavis) [inline Godeps/_workspace/src/github.com/tendermint/tendermint/vm/vm.go]: We keeping this as todo?

Comment (benjaminbollen) [inline Godeps/_workspace/src/github.com/tendermint/tendermint/vm/vm.go]: It is arguably redundant as value cannot be passed to another contract if it did not exist; and value is atomically swapped.  However, should a contract rely on `CALLVALUE` Im at this point not confident it would not be possible to produce bytecode that would put a malignent value on the stack and as such provide a hidden way to corrupt seemingly valid solidity.

File: Godeps/_workspace/src/github.com/tendermint/tendermint/vm/vm.go [hunk 4]
```diff
@@ -745,12 +769,21 @@ func (vm *VM) call(caller, callee *Account, code, input []byte, value int64, gas
...
 			if !HasPermission(vm.appState, callee, ptypes.Call) {
 				return nil, ErrPermission{"call"}
 			}
 			gasLimit := stack.Pop64()
-			addr, value := stack.Pop(), stack.Pop64()
+			addr := stack.Pop()
+			// NOTE: for DELEGATECALL value is preserved from the original
+			// caller, as such it is not stored on stack for the address
+			// to be called; for CALL and CALLCODE value is stored on stack
+			// and needs to be overwritten from the given value.
```
Comment (silasdavis) [inline Godeps/_workspace/src/github.com/tendermint/tendermint/vm/vm.go]: I found this a bit hard to parse, having read it a few times over. Here is a proposed rewording for clarity, but happy to see an even better version if this doesn't convey proper meaning:

```
			// NOTE: for DELEGATECALL value is preserved from the original
			// caller, as such it is not stored on stack for as an argument for
                        // DELEGATECALL and should not be popped. Instead the previous
                        // caller value is use.
                        // for CALL and CALLCODE value is stored on stack
			// and needs to overwrite the original caller value.
```

**Comment by silasdavis** [comment]
So other than my suggestion for a rewording of the comment describing the crucial implementation fact for DELEGATECALL, this looks good to me!

Great work Ben

File: Godeps/_workspace/src/github.com/tendermint/tendermint/vm/vm.go (hunk 4 above)
Comment (benjaminbollen) [inline Godeps/_workspace/src/github.com/tendermint/tendermint/vm/vm.go]: I was making these changes on develop; ok to just take the new comment on 0.12 develop? And leave it for hotfix here as is?
